diff --git a/node_modules/global-agent/dist/classes/Agent.js b/node_modules/global-agent/dist/classes/Agent.js
index ba8cc1a81..1e9bd4e8f 100644
--- a/node_modules/global-agent/dist/classes/Agent.js
+++ b/node_modules/global-agent/dist/classes/Agent.js
@@ -40,6 +40,18 @@ class Agent {
       requestUrl = this.protocol + '//' + (configuration.hostname || configuration.host) + (configuration.port === 80 || configuration.port === 443 ? '' : ':' + configuration.port) + request.path;
     }
 
+    // If a request should go to a local socket, proxying it through an HTTP
+    // server does not make sense as the information about the target socket
+    // will be lost and the proxy won't be able to correctly handle the request.
+    if (configuration.socketPath) {
+      log.trace({
+        destination: configuration.socketPath,
+      }, "not proxying request; destination is a socket");
+
+      this.fallbackAgent.addRequest(request, configuration);
+      return;
+    }
+
     if (!this.isProxyConfigured()) {
       log.trace({
         destination: requestUrl
